### ISSUE TITLE
bug(ErrorStateMatcher): Not able to configure globally ErrorStateMatcher

### ISSUE THREAD

**Issue by wis-ekubesh** [angular/components#20097]
#### Reproduction

StackBlitz link for reproduction: https://stackblitz.com/edit/components-issue-enshvf

Steps to reproduce:
1. Pay your attention that CoreModule has declaration for `ErrorStateMatcher` provider with class `ShowOnDirtyErrorStateMatcher`. CoreModule imported into AppModule in the same way as FeatureModule.
2. Try to focus/blur example "Favorite Food" input and note that `ShowOnDirtyErrorStateMatcher` was not applied.

In big enterprise applications we should have clean modular structure without any SharedModule or MaterialModule to be imported into each feature module.

MatInputModule has hardcoded `ErrorStateMatcher` provider and when we import that module into our feature modules – our global `ErrorStateMatcher` provider is overwritten.
 
I suppose this issue related with PR #18962, but I believe this issue describes another problem with `ErrorStateMatcher`
 
#### Expected Behavior

ErrorStateMatcher can be defined globally

#### Actual Behavior

ErrorStateMatcher overwrites with each MatInputModule import

#### Environment

- Angular: 9+
- CDK/Material: 9+

**Comment by crisbeto** [comment]
We should follow [the same pattern as in the `MatDateSelectionModel`](https://github.com/angular/components/blob/master/src/material/datepicker/date-selection-model.ts#L194) when providing the default `ErrorStateMatcher` so that any higher-level defaults are picked up.

**Comment by wis-ekubesh** [comment]
> We should follow [the same pattern as in the `MatDateSelectionModel`](https://github.com/angular/components/blob/master/src/material/datepicker/date-selection-model.ts#L194) (`SkipSelf` + `Optional`) when providing the default `ErrorStateMatcher` so that any higher-level defaults are picked up.

Do you need my help with fixing this?

**Comment by nicolae536** [comment]
Yes this would be very helpfull cause right now you cannot override the default you can pass it using inputs but that creates a lot of boilerplate

**Comment by ngehlert** [comment]
is this still an issue? any updates? does it make sense to create a MR for this?
